rust/vapid/src/lib.rs

```diff
@@ -0,0 +1,482 @@
+        // Return the private key as a raw bit array
+        let key = self.0.private_key().expect("Could not extract private key");
+        String::from_utf8(key.to_vec()).expect("Could not convert private key to string")
+    }
+
+    /// Derive a vapid key from the provided raw base64 public key string
+    fn derive_ec_public_point(bits: String) -> Key {
+        //Read a private key from a raw bit array
+        let group = Key::group();
+        let mut ctx = BigNumContext::new().unwrap();
+        let mut new_key = ec::EcKeyBuilder::new().unwrap();
+
+        let bytes: Vec<u8> = bits.into_bytes();
+        let secret = BigNum::from_slice(&bytes).unwrap();
+        let mut point = ec::EcPoint::new(&group).unwrap();
+        point.mul_generator(&group, &secret, &mut ctx).unwrap();
+        new_key.set_group(&group).expect("Could not set group");
+        new_key.set_public_key(&point).expect(
+            "Could not set public key",
+        );
+        new_key.set_private_key(&secret).expect(
+            "Could not set private key",
+        );
+        Key(new_key.build())
+    }
+
+    /// Convert the public key into a uncompressed, raw base64 string
+    pub fn to_public_raw(&self) -> String {
+        //Return the public key as a raw bit array
+        let mut ctx = BigNumContext::new().unwrap();
+        let group = Key::group();
+
+        let key = self.0.public_key().unwrap();
+        let keybytes = key.to_bytes(&group, ec::POINT_CONVERSION_UNCOMPRESSED, &mut ctx)
+            .unwrap();
+        base64::encode_config(&keybytes, base64::URL_SAFE_NO_PAD)
+    }
+
+    /// Read the public key from an uncompressed, raw base64 string
+    pub fn from_public_raw(bits: String) -> Result<Key, String> {
+        //Read a public key from a raw bit array
+        let b64bytes = &bits.into_bytes();
+        let bytes: Vec<u8> = base64::decode_config(b64bytes, base64::URL_SAFE_NO_PAD).unwrap();
+        let group = Key::group();
+        if bytes.len() != 65 || bytes[0] != 4 {
+            // It's not a properly tagged key.
+            return Err("Invalid key format for public key.".into());
+        }
+
+        let x = openssl::bn::BigNum::from_slice(&bytes[1..33]).expect("Invalid public key");
+        let y = openssl::bn::BigNum::from_slice(&bytes[33..]).expect("Invalid public key");
+        let mut new_key = match ec::EcKeyBuilder::new() {
+            Ok(key) => key,
+            Err(err) => return Err(format!("Could not generate key: {}", err)),
+        };
+        new_key.set_group(&group).expect("Could not set group");
+        new_key.set_public_key_affine_coordinates(&x, &y).expect(
+            "Invalid coordiates for public key",
+        );
+        Ok(Key(new_key.build()))
+    }
+}
+
+#[derive(Debug)]
+struct AuthElements {
+    t: Vec<String>,
+    k: String,
+}
+
+/// Parse the Authorization Header for useful things.
+fn parse_auth_token(auth_token: &mut String) -> Result<AuthElements, String> {
```

Comment:
Similar to above I think you can just take `auth_token: &str` here instead of a `&mut String` (I don't think this mutates the `auth_token` itself in the function)